packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,434 @@
+// @ts-check
+
+/// <reference types="ses"/>
+
+// This module is based on the `encodePassable.js` in `@agoric/store`,
+// which may migrate here. The main external difference is that
+// `encodePassable` goes directly to string, whereas `encodeToSmallcaps`
+// encodes to Smallcaps, a JSON-representable data structure, and leaves it to
+// the caller (`marshal.js`) to stringify it.
+
+import { passStyleOf } from './passStyleOf.js';
+
+import { ErrorHelper } from './helpers/error.js';
+import { makeTagged } from './makeTagged.js';
+import { getTag, hasOwnPropertyOf } from './helpers/passStyle-helpers.js';
+import {
+  assertPassableSymbol,
+  nameForPassableSymbol,
+  passableSymbolForName,
+} from './helpers/symbol.js';
+
+/** @typedef {import('./types.js').Passable} Passable */
+/** @typedef {import('./types.js').Remotable} Remotable */
+// @typedef {import('./types.js').SmallcapsEncoding} SmallcapsEncoding */
+// @typedef {import('./types.js').SmallcapsEncodingUnion} SmallcapsEncodingUnion */
+/** @typedef {any} SmallcapsEncoding */
+/** @typedef {any} SmallcapsEncodingUnion */
+
+const { ownKeys } = Reflect;
+const { isArray } = Array;
+const { is, fromEntries } = Object;
+const { details: X, quote: q } = assert;
+
+/**
+ * Smallcaps considers the characters between `!` (ascii code 33)
+ * and `-` (ascii code 45) to be special. These characters, in order, are
+ * `!"#$%&'()*+,-` Of these, smallcaps currently uses the following:
+ *
+ *  * `'` - escaped string
+ *  * `+` - non-negative bigint
+ *  * `-` - negative bigint
+ *  * `#` - manifest constant
+ *  * `%` - symbol
+ *  * `$` - remotable
+ *  * `&` - promise
+ *
+ * All other special characters (`!"()*,`) are reserved for future use.
+ *
+ * The menifest constants that smallcaps currently uses for values:
+ *  * `#undefined`
+ *  * `#NaN`
+ *  * `#Infinity`
+ *  * `#-Infinity`
+ *
+ * and for property names:
+ *  * `#tagged`
+ *  * `#error`
+ *
+ * All other encoded strings beginning with `#` are reserved for
+ * future use.
+ *
+ * @param {string} encodedStr
+ * @returns {boolean}
+ */
+const startsSpecial = encodedStr => {
+  if (encodedStr === '') {
+    return false;
+  }
+  const c = encodedStr[0];
+  // eslint-disable-next-line yoda
+  return '!' <= c && c <= '-';
```

Comment:
I believe `.charCodeAt(0)` and number comparison is faster than `[0]` and string comparison.

Some "old" thread about it, I'm sure there are others:
https://www.reddit.com/r/webdev/comments/e2eft5/til_that_charcodeat_is_faster_than_accessing_a/

Reply:
See conversation at https://github.com/endojs/endo/pull/1284#pullrequestreview-1106150594

I still don't want to do this unless measurements show that it is needed on XS. Resolving.

Reply:
I also second @mhofman's suggestion of `EXCLAMATION` and `DASH` constants from https://github.com/endojs/endo/pull/1284#discussion_r969959237

Reply:
Call and raise `const BANG = '!'.charCodeAt(0)`

Reply:
Measurements, thanks!

Within `startsSpecial`, done.